lib/logstash/outputs/graphite.rb

```diff
@@ -86,18 +93,48 @@ def register
       @metrics_format = DEFAULT_METRICS_FORMAT
     end
 
-    connect
+    setup_hosts
+  end
+
+  def setup_hosts
+    if @hosts && @hosts.size == 1 && @hosts[0] == "localhost:2003"
+      @hosts.replace(["%s:%s" % [@host, @port]])
+    end
+  end
+
+  # just for convenient mock
+  def socket(address, port)
+    return TCPSocket.new(address, port)
   end
 
-  def connect
-    # TODO(sissel): Test error cases. Catch exceptions. Find fortune and glory. Retire to yak farm.
+  def transmit(message)
+    numattempts = 0
+    hosts_to_try = @hosts.clone
+    host = hosts_to_try.delete hosts_to_try.sample
+
     begin
-      @socket = TCPSocket.new(@host, @port)
-    rescue Errno::ECONNREFUSED => e
-      @logger.warn("Connection refused to graphite server, sleeping...", :host => @host, :port => @port)
-      sleep(@reconnect_interval)
-      retry
+      address, _, port = host.rpartition(":")
+      @logger.debug? && @logger.debug("Trying to send metrics to", :address => address, :port => port)
+      socket(address, port).puts(message)
+    rescue Exception => e
+      @logger.debug? && @logger.debug("Suffering from", :e => e.message)
```

Comment:
`Suffering` is probably not the right word. Also, catching Exception isn't the right thing, you probably want StandardError, if the case is that you want a general catch-all. `Exception` will catch things like SyntaxError or NoMemoryError that you probably dont' want to be catching here.